src/wrapper/jnienv.rs

```diff
@@ -349,11 +349,15 @@ impl<'a> JNIEnv<'a> {
 
     /// Create a new local ref to an object.
     ///
-    /// Note that the object passed to this is *already* a local ref. This
-    /// creates yet another reference to it, which is most likely not what you
-    /// want.
-    pub fn new_local_ref<T>(&self, obj: JObject<'a>) -> Result<JObject<'a>> {
-        let local: JObject = jni_unchecked!(self.internal, NewLocalRef, obj.into_inner()).into();
+    /// This is useful for creating a local reference from a [global reference][GlobalRef] or
+    /// creating a local reference in a different [local reference frame][JNIEnv::with_local_frame]
+    /// than the original.
+    pub fn new_local_ref<'b, O>(&self, obj: O) -> Result<JObject<'a>>
+    where
+        O: Into<JObject<'b>>,
```

Comment:
>Wouldn't it make sense for `new_local_ref()` to return `JObject<'b>`?

>I mean, why do you wrap the lifetime-converted `jobject` into a `JObject<'a>`? So that the wrapper has the same lifetime than the original `JNIEnv`, I guess. But, doesn't that defeat this whole `NewLocalRef` business?

The idea is to create a new local reference that's independent of the original one, with a *different* lifetime than the original, that will still be valid even if the original is dropped. `'a` is the lifetime of the new reference and `'b` is the lifetime of the original reference.

Maybe it would help to read `'a` as `'new` and `'b` as `'original`? I know the lifetime names are kind of confusing, but I can't change `'a` here (it's part of the `impl` block), and having the two lifetimes be named `'a` and `'original` would be inconsistent.

The JNI function `NewLocalRef` creates a reference in the current local reference frame, regardless of whether the original reference belongs to the same local reference frame, a different one, or is a global reference. In Rust terms, it accepts input with any valid lifetime and produces a clone with its own lifetime. This PR changes `new_local_ref` to reflect that.

>Also, explaining / detailing your use case … would be good.

The main use case is upgrading weak references to strong references. It's used [here](https://github.com/jni-rs/jni-rs/blob/26bb8b43d8f8c1e59bc24c824cb8a811551ed129/src/wrapper/objects/weak_ref.rs#L77) in PR #304.

The object pointed to by a weak reference can be garbage collected (making the reference effectively `null`) at any moment. In order to use a weak reference without a possible `NullPointerException`, you have to create a new strong reference to whatever object the weak reference points to, then check if the new strong reference is `null`.

There are two ways to do that. One of them is `new_local_ref`, which accepts any kind of reference and creates a strong local reference to the same object. But global references (including weak ones) always have a different lifetime than any `JNIEnv`, so in order to use `new_local_ref` this way, it has to accept a reference with a lifetime different from that of the `JNIEnv` it's called on. Hence issue #301 and this PR.

Another use case is when you have a strong global reference and you can't prevent it from being dropped before you're finished with it. In that case, you can create a local reference using `new_local_ref`, which is guaranteed to remain valid for the duration of the current local reference frame, regardless of what later happens to the original global reference. Here's an example of some error-handling code in one of my projects that does this:

```rust
// Some preceding code has resulted in an error, caused either by a Java
// exception or something going wrong in Rust code. Either way, we need to
// return a Java `Throwable` from this function.

let update_exception: JObject = match update_error {
	UpdateError::Exception(exception) => {
		// The error was caused by a Java exception.

		// Here, `exception` is a `GlobalRef` pointing to a Java `Throwable`. It
		// will be dropped at the end of this `match` arm. We'll use
		// `new_local_ref` to create a local reference that will outlive the
		// `GlobalRef`.

		env.new_local_ref(exception.as_obj())?
	}

	UpdateError::Other(error) => {
		// The error was caused by something that happened in Rust code. Create a
		// new `java.lang.Error` to represent it.

		env.new_object(
			"java/lang/Error",
			"(Ljava/lang/String;)V",
			&[
				env.new_string(error.to_string())?.into(),
			],
		)?
	}
};
```

>Also, … perhaps adding some tests to check these lifetime bounds, would be good.

I don't think that's possible. There's no way to check whether a reference is valid. Calling any JNI function on an invalid reference results in either undefined behavior (without `-Xcheck:jni`) or [process abort (with `-Xcheck:jni`)](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/clopts002.html#CHDDEGBI).

This rewritten `new_local_ref` is *used* by PR #304, which [does have tests](https://github.com/jni-rs/jni-rs/blob/26bb8b43d8f8c1e59bc24c824cb8a811551ed129/tests/jni_weak_refs.rs), but that's only an indirect test of the `new_local_ref` lifetime change.

Reply:
Thanks for the detailed explanation.

This makes sense to me. I also think it's a good idea to merge this fix for `new_local_ref` first, and then rebase #304 on top of it, and continue the discussion about your new impls there.

As I said, I'm not an expert on this, at all. Let's wait for @stanislav-tkach and @dmitry-timofeev feedback / comments.

Reply:
Also, renaming this PR to something like "Fix new_local_ref lifetimes" or "Remove new_local_ref lifetime restrictions" would be good.

Reply:
I am convinced by these arguments. :upside_down_face:

Reply:
Perhaps it would make sense to add some of these explanations to the function documentation?